### PR TITLE
[wdspec] Add tests for permissions

### LINKED ISSUE #43305: Umbrella bug for WebDriver BiDi Browser Permissions API tests
Spec: https://github.com/w3c/permissions/pull/425 (as an extension of: https://w3c.github.io/webdriver-bidi/)

cc @OrKoN @Lightning00Blade FYI

AI: Write wdspec tests for this new extension module.

### PR DESCRIPTION
**Pull request by OrKoN**
Spec: https://www.w3.org/TR/permissions/#automation-webdriver-bidi
Implementation chromium-bidi: https://github.com/GoogleChromeLabs/chromium-bidi/pull/1645
Closes #43305

### CODE REVIEW

File: tools/webdriver/webdriver/bidi/modules/permissions.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    @command
+    def end(self) -> Mapping[str, Any]:
+        return {}
+
+    @end.result
+    async def _end(self, result: Mapping[str, Any]) -> Any:
+        if self.session.transport:
+            await self.session.transport.wait_closed()
+
+        return result
```
Comment (sadym-chromium) [inline tools/webdriver/webdriver/bidi/modules/permissions.py:16]: I'm not sure why this is required.

File: tools/webdriver/webdriver/bidi/modules/permissions.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    @command
+    def set_permission(self,
+                    descriptor: Optional[Mapping[str, Any]] = None,
+                    state: Optional[str] = None,
+                    origin: Optional[str] = None) -> Mapping[str, Any]:
+        params: MutableMapping[str, Any] = {
+            "descriptor": descriptor,
+            "state": state,
+            "origin": origin
+        }
```
Comment (sadym-chromium) [inline tools/webdriver/webdriver/bidi/modules/permissions.py:27]: I'd assume you want to emit not provided parameters, not set them to `None` (`null`).

Comment (sadym-chromium) [reply tools/webdriver/webdriver/bidi/modules/permissions.py:27]: You can make it like this: 
https://github.com/web-platform-tests/wpt/blob/53d5a05d1cd02abe604721e5baf53e3ced7a960e/tools/webdriver/webdriver/bidi/modules/storage.py#L22C1-L44C36

Comment (OrKoN) [reply tools/webdriver/webdriver/bidi/modules/permissions.py:27]: I believe undefined should be done using https://github.com/web-platform-tests/wpt/blob/master/webdriver/tests/bidi/browsing_context/set_viewport/viewport.py#L2 Let me add that

File: tools/webdriver/webdriver/bidi/modules/permissions.py (hunk 1 above)
Comment (OrKoN) [reply tools/webdriver/webdriver/bidi/modules/permissions.py]: Removed. I think it was from another module.

File: tools/webdriver/webdriver/bidi/modules/permissions.py [hunk 3]
```diff
@@ -0,0 +1,17 @@
+from typing import Any, Optional, Mapping, MutableMapping
+
+from ._module import BidiModule, command
+
+
+class Permissions(BidiModule):
+    @command
+    def set_permission(self,
+                    descriptor: Optional[Mapping[str, Any]] = None,
```
Comment (sadym-chromium) [inline tools/webdriver/webdriver/bidi/modules/permissions.py:9]: As discussed offline, `None` should be replaced with `Undefined`

File: webdriver/tests/bidi/permissions/set_permission/invalid.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    with pytest.raises(error.InvalidArgumentException):
+      await bidi_session.permissions.set_permission(
+         descriptor={"name": "geolocation"},
+         state=state,
+         origin="https://example.com",
+      )
+
+
+@pytest.mark.parametrize("origin", [False, 42, {}, [], None, UNDEFINED])
+async def test_params_origin_invalid_type(bidi_session, origin):
```
Comment (sadym-chromium) [inline webdriver/tests/bidi/permissions/set_permission/invalid.py:48]: What about unknown origin? Should it set permissions successfully?

Comment (OrKoN) [reply webdriver/tests/bidi/permissions/set_permission/invalid.py:48]: Yes (no error), according to the current spec. Let me add a test.

Comment (OrKoN) [reply webdriver/tests/bidi/permissions/set_permission/invalid.py:50]: Updated

File: webdriver/tests/bidi/permissions/set_permission/invalid.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+      )
+
+@pytest.mark.parametrize("origin", ['UNKNOWN', ''])
+async def test_params_origin_invalid_value(bidi_session, new_tab, origin):
+    await bidi_session.permissions.set_permission(
+        descriptor={"name": "geolocation"},
+        state="granted",
+        origin=origin,
+    )
+    assert await get_permission_state(bidi_session, new_tab,  "geolocation") == "prompt"
```
[the remark this answers is not in the document]
Comment (OrKoN) [reply webdriver/tests/bidi/permissions/set_permission/invalid.py:65]: well, it's still an invalid origin

**Review by sadym-chromium — APPROVED** [review]

File: webdriver/tests/bidi/permissions/set_permission/set_permission.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+        url=test_url,
+        wait="complete",
+    )
+
+    # See https://github.com/w3c/permissions/issues/437.
+    assert await get_permission_state(bidi_session, new_context,  "geolocation") == "granted"
+
+
+@pytest.mark.parametrize("origin", ['UNKNOWN', ''])
+async def test_params_origin_invalid_value(bidi_session, new_tab, origin):
```
Comment (sadym-chromium) [inline webdriver/tests/bidi/permissions/set_permission/set_permission.py:99]:
```suggestion
async def test_params_origin_unknown_value(bidi_session, new_tab, origin):
```

**Review by sadym-chromium — APPROVED** [review]

**Review by sadym-chromium — APPROVED** [review]

**Comment by OrKoN** [comment]
@whimboo the spec is here https://www.w3.org/TR/permissions/#automation-webdriver-bidi (same as webdriver/tests/classic/permissions/set.py)

**Comment by whimboo** [comment]
Please note that you are adding a `permissions` module including commands for WebDriver BiDi. As long as this is not available in the BiDi spec the tests utilizing those commands have to be marked as tentative.

**Comment by jgraham** [comment]
I don't think they need to be marked tentative, but I wonder if we should have a directory like `webdriver/bidi/external/permissions` or something. wpt limitations mean that all WebDriver tests have to be in `webdriver/tests`, but making it clear which ones are related to the base WebDriver spec and which are related to other specifications seems like it would be helpful.

**Comment by OrKoN** [comment]
https://github.com/web-platform-tests/wpt/pull/44045

**Pull request merged**